mmaprototype: dedup constraint conj

wenyihu6 opened the pull request:
Epic: CRDB-55052 
Resolves: #159211
Release note: none

---
**mmaprototype: add more edge cases for dup constraints**

This commit adds additional duplicate-constraint edge cases to demonstrate the
current incorrect output. These will be addressed in upcoming commits.

---
**mmaprototype: add dedupConstraints**

This commit introduces dedupConstraints, which the normalization logic will use
to merge duplicate constraint conjunctions.

---
**mmaprototype: dedup constraint conj**

Previously, identical constraints were left as duplicates and processed
separately, leading to invalid normalization results. This commit combines
duplicate constraints by summing their NumReplicas.

Deduplication is needed at multiple stages:
- During basic normalization (normalizeConstraints): to handle user-specified
  duplicates in the input.
- During voter constraint normalization: splitting constraints to align with
  voter constraints can produce duplicates.
- During empty constraint normalization: synthesizing empty constraints to fill
  gaps can create duplicates with user-specified empty constraints.

---
**mmaprototype: panic instead for dup empty constraints**

Previously, we fixed duplicate constraint conjunctions by deduping them. With
that in place, this commit now makes duplicate empty constraints a hard error
and panic on them because they indicate a programming mistake.

wenyihu6 committed:
mmaprototype: add more edge cases for dup constraints

This commit adds additional duplicate-constraint edge cases to demonstrate the
current incorrect output. These will be addressed in upcoming commits.

wenyihu6 committed:
mmaprototype: add dedupConstraints

This commit introduces dedupConstraints, which the normalization logic will use
to merge duplicate constraint conjunctions.

wenyihu6 committed:
mmaprototype: dedup constraint conj

Previously, identical constraints were left as duplicates and processed
separately, leading to invalid normalization results. This commit combines
duplicate constraints by summing their NumReplicas.

Deduplication is needed at multiple stages:
- During basic normalization (normalizeConstraints): to handle user-specified
  duplicates in the input.
- During voter constraint normalization: splitting constraints to align with
  voter constraints can produce duplicates.
- During empty constraint normalization: synthesizing empty constraints to fill
  gaps can create duplicates with user-specified empty constraints.

wenyihu6 committed:
mmaprototype: panic instead for dup empty constraints

Previously, we fixed duplicate constraint conjunctions by deduping them. With
that in place, this commit now makes duplicate empty constraints a hard error
and panic on them because they indicate a programming mistake.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159210)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> @sumeerbhola reviewed 1 of 1 files at r1, 3 of 3 files at r2, 3 of 3 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 654 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAf4h0AHLWGbgjCBBF:-OgAf4h15Mpy7jk5DBtS:by1080u) ([raw file](https://github.com/cockroachdb/cockroach/blob/00e35fc0b16223da6b8a30d72652de47a6fd9cd1/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L654)):*
> ```Go
> 		if len(conf.voterConstraints[i].constraints) == 0 {
> 			if emptyVoterConstraintIndex != -1 {
> 				log.Fatalf("multiple empty voter constraints: %v and %v",
> ```

On second thought, I think we should not panic/fatal anywhere in span config normalization code, and simply return errors.
And if we fail even basic normalization, we should not panic either. Since we can't do any work on a range which fails even basic normalization, I suppose we'll just have to ignore the range when rebalancing etc., with loud error messages and metrics.

The reasoning is that if a new SpanConfig triggers a bug in our code, it can crash all the nodes in a cluster at the same time. And these crashes will keep repeating. We should adhere to the principle of no fatal errors on user provided input, even if the fault is ours.

All of the above can happen in a followup PR, but we should prioritize it.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAcssy7dmLVvg38q5y:-OgAcssy7dmLVvg38q5z:bwt7q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L31)):*
> ```
>  num-replicas=3 num-voters=3
>  constraints:
>    :1
> ```

The comment says :3, but here it is :1
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAd1Rm9E7Ct2ErxDZk:-OgAd1Rm9E7Ct2ErxDZl:be2ys07) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L73)):*
> ```
> 
> # Voter constraints - duplicates dedupd. Normalization also narrows the
> # all-replica constraints based on voter constraints,
> ```

I don't see any narrowing of the replica constraints in the output.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAd9pCCN-Is1lKc1MS:-OgAd9pCCN-Is1lKc1MT:bd4mrzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L92)):*
> ```
>  num-replicas=5 num-voters=3
>  voter-constraints:
>    :1
> ```

This shows :1 which contradicts the comment.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint_test.go` line 962 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAbwJj4pWYRA_R7WQQ:-OgAbwJj4pWYRA_R7WQR:bo28ieu) ([raw file](https://github.com/cockroachdb/cockroach/blob/00e35fc0b16223da6b8a30d72652de47a6fd9cd1/pkg/kv/kvserver/allocator/mmaprototype/constraint_test.go#L962)):*
> ```Go
> 			case "dedup":
> 				// Parse span config.
> 				conf := parseSpanConfig(t, d)
> ```

I don't understand why this test exists, given it is starting with `parseSpanConfig`. I assumed based on the name that it was a test for a single set of constraints conjunction, for which this would test `dedupConstraints` directly. If we are going to test `dedupConstraints` via a full SpanConfig (which is fine), then why can we not simply use `TestNormalizedSpanConfig` and put any new test cases in another testdata file. Because eventually that function will be calling `dedupConstraints`.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 774 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAe-aI9N-dOHQMsfQY:-OgAe-aJC0Gchlruvhpu:bnd8maq) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L774)):*
> ```Go
> 		vc = append(vc, ncEnv.voterConstraints[i].internedConstraintsConjunction)
> 	}
> 	return dedupConstraints(vc)
> ```

I see now why dedupConstraints was also filtering the 0 numReplicas case.
If you want to do that, it should be mentioned in the comment, and perhaps rename it to dedupAndFilterConstraints.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAInD_-0mIq3qJRogp:-OgAInD_-0mIq3qJRogq:b-wnq5jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L378)):*
> ```Go
> // constraints: [+region=us-west-1]: 1, [+region=us-west-1]: 1, []: 3, []: 1
> // result: [+region=us-west-1]: 2, []: 4
> func dedupConstraints(
> ```

I suppose this will be called multiple times during normalization of a SpanConfig. In which case it is better to reuse the map, by passing in a scratch map as a parameter.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 389 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAIRLbE3eaNwvipsTq:-OgAIRLbE3eaNwvipsTr:b-j1nja6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L389)):*
> ```Go
> 	result := make([]internedConstraintsConjunction, 0, len(constraints))
> 	for i := range constraints {
> 		if constraints[i].numReplicas == 0 {
> ```

I think we should just declare an invariant in `internedConstraintsConjunction` that `numReplicas > 0`. We are always careful to initialize the struct with a non-zero value.
Then we don't need to do anything special here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAQiUdBGDPEdjQxpuv:-OgAQiUdBGDPEdjQxpuw:b-etz0lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L413)):*
> ```Go
> 			seen[h] = []int{len(result)}
> 		}
> 		result = append(result, constraints[i])
> ```

If we wanted to do this without a hash table and without allocating a slice for each hashtable entry, we could do something like (does one slice allocation):

```
func (cc constraintsConj) cmp(b constraintsConj) int {
	n := min(len(cc), len(b))
	for i := 0; i < n; i++ {
		cmp := cc[i].cmp(b[i])
		if cmp != 0 {
			return cmp
		}
	}
	if n < len(cc) {
		return +1
	}
	if n < len(b) {
		return -1
	}
	return 0
}

func dedupConstraints(
	constraints []internedConstraintsConjunction,
) []internedConstraintsConjunction {
	n := len(constraints)
	// indices represents the positions in the original constraints.
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	// Sort the indices such that equal constraints get grouped and within a
	// sequence of equal constraints, the smallest index sorts first. The latter
	// allows us to preserve the original ordering after de-duplication.
	slices.SortFunc(indices, func(i, j int) int {
		return cmp.Or(constraints[i].constraints.cmp(constraints[j].constraints),
			cmp.Compare(i, j))
	})
	// i represents iteration over the indices slice.
	i := 1
	// j is the index into the updated indices slice, where the value
	// is the index of the original constraint that is preserved.
	j := 0
	// Say the original indices slice is the following, where the parentheses
	// represent equal constraints: (3, 5), (1, 4), (0, 2). At the end, we will
	// have 3, 1, 0 in the slice.
	for ; i < n; i++ {
		comp := constraints[indices[i-1]].constraints.cmp(constraints[indices[i]].constraints)
		if comp == 0 {
			constraints[indices[i-1]].numReplicas += constraints[indices[i]].numReplicas
		} else {
			j++
			indices[j] = indices[i]
		}
	}
	indices = indices[:j+1]
	// Sort these indices in increasing order.
	slices.Sort(indices)
	for i := range indices {
		constraints[i] = constraints[indices[i]]
	}
	return constraints[:len(indices)]
}
```
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAeRik8TBkK75sJEVD:-OgAeRik8TBkK75sJEVE:bjjcert) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L238)):*
> ```
> 
> # Test error case: multiple empty constraints (via normalization adding empty
> # constraint).
> ```

comment is stale
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 263 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAeVaC3GL4kCdnE8s-:-OgAeVaC3GL4kCdnE8s0:bjjcert) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L263)):*
> ```
> 	[idx=1][voter=+region=us-west:1] [all=:8] rel=strictSubset
> 
> # Test error case: multiple explicit empty constraints.
> ```

comment is stale


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint_test.go` line 962 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAbwJj4pWYRA_R7WQQ:-OgFdfAs9LDtVy8bd9SP:b-sl4pp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/00e35fc0b16223da6b8a30d72652de47a6fd9cd1/pkg/kv/kvserver/allocator/mmaprototype/constraint_test.go#L962)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand why this test exists, given it is starting with `parseSpanConfig`. I assumed based on the name that it was a test for a single set of constraints conjunction, for which this would test `dedupConstraints` directly. If we are going to test `dedupConstraints` via a full SpanConfig (which is fine), then why can we not simply use `TestNormalizedSpanConfig` and put any new test cases in another testdata file. Because eventually that function will be calling `dedupConstraints`.
</blockquote></details>

I think my comments in the test are outdated. I had initially followed your suggestion using `makeNormalizedSpanConfigWithObserver`, but I later thought I prefer a unit test focused solely on `dedupConstraints`. That’s still my preference, so I’ll update `TestDedupConstraints` to test a single set of constraints, which means I won’t be able to use helper functions like `parseSpanConfig` or `printSpanConfig`. Lmk if our preference is to use `makeNormalizedSpanConfigWithObserver` though. 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: rename dedup* to dedupAndFilterConstraints

wenyihu6 committed:
mmaprototype: use sort based de-duplication for dedupAndFilterConstraints

This commit updates dedupAndFilterConstraints to use sort based approach to
perform de-duplication.

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint_test.go` line 962 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAbwJj4pWYRA_R7WQQ:-OgNtYVZD1ISKXyimmtI:b-8dyhsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/00e35fc0b16223da6b8a30d72652de47a6fd9cd1/pkg/kv/kvserver/allocator/mmaprototype/constraint_test.go#L962)):*
> I later thought I prefer a unit test focused solely on `dedupConstraints`. That’s still my preference, so I’ll update `TestDedupConstraints` to test a single set of constraints

That's fine.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
The constraint.go changes look good.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-:-OgNuxRT9BZupQbaBtJp:b-9i09k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuENi9bnhPYtucfys:-OgNuENjC7EV1eHWULFH:b-noss5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L137)):*
> ```Go
> // does not imply that +region=a is stricter than -region=b. They simply mean
> // that they are two distinct constraints.
> func (ic internedConstraint) less(b internedConstraint) bool {
> ```

let's remove `less` now that we have `cmp`.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 160 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuODu3cpV26wVvFXH:-OgNuODu3cpV26wVvFXI:b1y80mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L160)):*
> ```Go
> // for optimization, and must be accompanied by a full equality check for
> // correctness.
> func (cc constraintsConj) hash() uint64 {
> ```

do we need this hash?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
The constraint.go changes look good.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-:-OgNuxRT9BZupQbaBtJp:b-9i09k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuENi9bnhPYtucfys:-OgNuENjC7EV1eHWULFH:b-noss5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L137)):*
> ```Go
> // does not imply that +region=a is stricter than -region=b. They simply mean
> // that they are two distinct constraints.
> func (ic internedConstraint) less(b internedConstraint) bool {
> ```

let's remove `less` now that we have `cmp`.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 160 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuODu3cpV26wVvFXH:-OgNuODu3cpV26wVvFXI:b1y80mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L160)):*
> ```Go
> // for optimization, and must be accompanied by a full equality check for
> // correctness.
> func (cc constraintsConj) hash() uint64 {
> ```

do we need this hash?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: improve TestDedupAndFilterConstraints

Previously, TestDedupAndFilterConstraints took span config, even though it
was only used for dedupAndFilterConstraints, which itself takes
[]internedConstraintsConjunction. This was confusing. This change updates
TestDedupAndFilterConstraints to take []internedConstraintsConjunction directly,
making the test’s input clearer.

wenyihu6 committed:
mmaprototype: change internedConstraint.less to cmp < 1

Previously, internedConstraint.less was used to compare two internedConstraint
values. Now that internedConstraint.cmp provides direct comparison, this change
updates all uses of the less helper to rely on cmp, comparing against -1
instead.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAInD_-0mIq3qJRogp:-OgXRpn3-L1PciXGpjzb:b8jvo76) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L378)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I suppose this will be called multiple times during normalization of a SpanConfig. In which case it is better to reuse the map, by passing in a scratch map as a parameter.
</blockquote></details>

No longer using a map. We still have one allocation from the indices that we could pass around as a parameter if we would want. I leave this as is for now. I will add a TODO.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 389 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAIRLbE3eaNwvipsTq:-OgXS4659krerC8ZDW3r:b-klpdcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L389)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think we should just declare an invariant in `internedConstraintsConjunction` that `numReplicas > 0`. We are always careful to initialize the struct with a non-zero value.
Then we don't need to do anything special here.
</blockquote></details>

Based on your later comment https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAe-aI9N-dOHQMsfQY, I think assume this makes sense and we want to keep it? It is because when we normalize voter or empty constraints, we can end up with zero replicas for a constraint. I renamed and improved the comments around here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAQiUdBGDPEdjQxpuv:-OgXT7_h0x6k7yo1o5Fq:bc5a5dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L413)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

If we wanted to do this without a hash table and without allocating a slice for each hashtable entry, we could do something like (does one slice allocation):

```
func (cc constraintsConj) cmp(b constraintsConj) int {
	n := min(len(cc), len(b))
	for i := 0; i < n; i++ {
		cmp := cc[i].cmp(b[i])
		if cmp != 0 {
			return cmp
		}
	}
	if n < len(cc) {
		return +1
	}
	if n < len(b) {
		return -1
	}
	return 0
}

func dedupConstraints(
	constraints []internedConstraintsConjunction,
) []internedConstraintsConjunction {
	n := len(constraints)
	// indices represents the positions in the original constraints.
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	// Sort the indices such that equal constraints get grouped and within a
	// sequence of equal constraints, the smallest index sorts first. The latter
	// allows us to preserve the original ordering after de-duplication.
	slices.SortFunc(indices, func(i, j int) int {
		return cmp.Or(constraints[i].constraints.cmp(constraints[j].constraints),
			cmp.Compare(i, j))
	})
	// i represents iteration over the indices slice.
	i := 1
	// j is the index into the updated indices slice, where the value
	// is the index of the original constraint that is preserved.
	j := 0
	// Say the original indices slice is the following, where the parentheses
	// represent equal constraints: (3, 5), (1, 4), (0, 2). At the end, we will
	// have 3, 1, 0 in the slice.
	for ; i < n; i++ {
		comp := constraints[indices[i-1]].constraints.cmp(constraints[indices[i]].constraints)
		if comp == 0 {
			constraints[indices[i-1]].numReplicas += constraints[indices[i]].numReplicas
		} else {
			j++
			indices[j] = indices[i]
		}
	}
	indices = indices[:j+1]
	// Sort these indices in increasing order.
	slices.Sort(indices)
	for i := range indices {
		constraints[i] = constraints[indices[i]]
	}
	return constraints[:len(indices)]
}
```
</blockquote></details>

Done. One note in case this wasn’t noted: I slightly changed this part of the code to handle cases where multiple constraints merge into one.

```
comp := constraints[indices[i-1]].constraints.cmp(constraints[indices[i]].constraints)
     if comp == 0 {
	   constraints[indices[i-1]].numReplicas += constraints[indices[i]].numReplicas
     }
```
to 
```
if constraints[indices[j]].constraints.cmp(constraints[indices[i]].constraints) == 0 {
		constraints[indices[j]].numReplicas += constraints[indices[i]].numReplicas
}
```
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 774 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAe-aI9N-dOHQMsfQY:-OgXUuAMBB-y2d0mdXt6:b-219cgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L774)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I see now why dedupConstraints was also filtering the 0 numReplicas case.
If you want to do that, it should be mentioned in the comment, and perhaps rename it to dedupAndFilterConstraints.
</blockquote></details>

Added comments and renamed. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 654 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAf4h0AHLWGbgjCBBF:-OgXUfev8_rE8jIVf39D:bi7qf1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/00e35fc0b16223da6b8a30d72652de47a6fd9cd1/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L654)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

On second thought, I think we should not panic/fatal anywhere in span config normalization code, and simply return errors.
And if we fail even basic normalization, we should not panic either. Since we can't do any work on a range which fails even basic normalization, I suppose we'll just have to ignore the range when rebalancing etc., with loud error messages and metrics.

The reasoning is that if a new SpanConfig triggers a bug in our code, it can crash all the nodes in a cluster at the same time. And these crashes will keep repeating. We should adhere to the principle of no fatal errors on user provided input, even if the fault is ours.

All of the above can happen in a followup PR, but we should prioritize it.
</blockquote></details>

Ack, added to ohttps://github.com/cockroachdb/cockroach/issues/158558. We can check in on the priority during mma sync. 

Updated: chatted more - I will file a new issue on this. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuENi9bnhPYtucfys:-OgXQHMRBPc82qe0pzbE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L137)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

let's remove `less` now that we have `cmp`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 160 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuODu3cpV26wVvFXH:-OgXQI2BBTWIUYZET2bL:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L160)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

do we need this hash?
</blockquote></details>

Removed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAeRik8TBkK75sJEVD:-OgXYFAV5cE8ZCVrkDHs:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L238)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

comment is stale
</blockquote></details>

Changed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 263 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAeVaC3GL4kCdnE8s-:-OgXXq1fCOXP1RSNMPt6:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L263)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

comment is stale
</blockquote></details>

Changed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAcssy7dmLVvg38q5y:-OgXUycB-JN0dJSr6ahP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L31)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The comment says :3, but here it is :1
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAd1Rm9E7Ct2ErxDZk:-OgXUwm17LBCs-eb1Amo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see any narrowing of the replica constraints in the output.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAd9pCCN-Is1lKc1MS:-OgXUyAoDU9p9TBd03CR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L92)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This shows :1 which contradicts the comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Filed https://github.com/cockroachdb/cockroach/issues/159531.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 1 of 4 files at r5, 2 of 2 files at r7, 1 of 2 files at r8, 1 of 1 files at r10, 1 of 1 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-:-OgYJUYYB9_eQqWi3oMr:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 207 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgYIPVD3IodFoO45AnJ:-OgYIPVD3IodFoO45AnK:bxcabwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/93b27579e52454aaedf9d46443ed7a7f3f2b8940/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L207)):*
> ```
> 	[idx=0][voter=+zone=us-west-a,+region=us-west:3] [all=+zone=us-west-b,+region=us-west:5] rel=nonIntersecting
> 
> # Test error case: multiple empty voter constraints are combined.
> ```

nit: this is no longer an "error case"


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159210)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAInD_-0mIq3qJRogp:-OgXRpn3-L1PciXGpjzb:b8jvo76) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L378)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I suppose this will be called multiple times during normalization of a SpanConfig. In which case it is better to reuse the map, by passing in a scratch map as a parameter.
</blockquote></details>

No longer using a map. We still have one allocation from the indices that we could pass around as a parameter if we would want. I leave this as is for now. I will add a TODO.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 389 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAIRLbE3eaNwvipsTq:-OgXS4659krerC8ZDW3r:b-klpdcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L389)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think we should just declare an invariant in `internedConstraintsConjunction` that `numReplicas > 0`. We are always careful to initialize the struct with a non-zero value.
Then we don't need to do anything special here.
</blockquote></details>

Based on your later comment https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAe-aI9N-dOHQMsfQY, I think assume this makes sense and we want to keep it? It is because when we normalize voter or empty constraints, we can end up with zero replicas for a constraint. I renamed and improved the comments around here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAQiUdBGDPEdjQxpuv:-OgXT7_h0x6k7yo1o5Fq:bc5a5dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L413)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

If we wanted to do this without a hash table and without allocating a slice for each hashtable entry, we could do something like (does one slice allocation):

```
func (cc constraintsConj) cmp(b constraintsConj) int {
	n := min(len(cc), len(b))
	for i := 0; i < n; i++ {
		cmp := cc[i].cmp(b[i])
		if cmp != 0 {
			return cmp
		}
	}
	if n < len(cc) {
		return +1
	}
	if n < len(b) {
		return -1
	}
	return 0
}

func dedupConstraints(
	constraints []internedConstraintsConjunction,
) []internedConstraintsConjunction {
	n := len(constraints)
	// indices represents the positions in the original constraints.
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	// Sort the indices such that equal constraints get grouped and within a
	// sequence of equal constraints, the smallest index sorts first. The latter
	// allows us to preserve the original ordering after de-duplication.
	slices.SortFunc(indices, func(i, j int) int {
		return cmp.Or(constraints[i].constraints.cmp(constraints[j].constraints),
			cmp.Compare(i, j))
	})
	// i represents iteration over the indices slice.
	i := 1
	// j is the index into the updated indices slice, where the value
	// is the index of the original constraint that is preserved.
	j := 0
	// Say the original indices slice is the following, where the parentheses
	// represent equal constraints: (3, 5), (1, 4), (0, 2). At the end, we will
	// have 3, 1, 0 in the slice.
	for ; i < n; i++ {
		comp := constraints[indices[i-1]].constraints.cmp(constraints[indices[i]].constraints)
		if comp == 0 {
			constraints[indices[i-1]].numReplicas += constraints[indices[i]].numReplicas
		} else {
			j++
			indices[j] = indices[i]
		}
	}
	indices = indices[:j+1]
	// Sort these indices in increasing order.
	slices.Sort(indices)
	for i := range indices {
		constraints[i] = constraints[indices[i]]
	}
	return constraints[:len(indices)]
}
```
</blockquote></details>

Done. One note in case this wasn’t noted: I slightly changed this part of the code to handle cases where multiple constraints merge into one.

```
comp := constraints[indices[i-1]].constraints.cmp(constraints[indices[i]].constraints)
     if comp == 0 {
	   constraints[indices[i-1]].numReplicas += constraints[indices[i]].numReplicas
     }
```
to 
```
if constraints[indices[j]].constraints.cmp(constraints[indices[i]].constraints) == 0 {
		constraints[indices[j]].numReplicas += constraints[indices[i]].numReplicas
}
```
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 774 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAe-aI9N-dOHQMsfQY:-OgXUuAMBB-y2d0mdXt6:b-219cgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L774)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I see now why dedupConstraints was also filtering the 0 numReplicas case.
If you want to do that, it should be mentioned in the comment, and perhaps rename it to dedupAndFilterConstraints.
</blockquote></details>

Added comments and renamed. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 654 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAf4h0AHLWGbgjCBBF:-OgXUfev8_rE8jIVf39D:bi7qf1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/00e35fc0b16223da6b8a30d72652de47a6fd9cd1/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L654)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

On second thought, I think we should not panic/fatal anywhere in span config normalization code, and simply return errors.
And if we fail even basic normalization, we should not panic either. Since we can't do any work on a range which fails even basic normalization, I suppose we'll just have to ignore the range when rebalancing etc., with loud error messages and metrics.

The reasoning is that if a new SpanConfig triggers a bug in our code, it can crash all the nodes in a cluster at the same time. And these crashes will keep repeating. We should adhere to the principle of no fatal errors on user provided input, even if the fault is ours.

All of the above can happen in a followup PR, but we should prioritize it.
</blockquote></details>

Ack, added to ohttps://github.com/cockroachdb/cockroach/issues/158558. We can check in on the priority during mma sync. 

Updated: chatted more - I will file a new issue on this. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuENi9bnhPYtucfys:-OgXQHMRBPc82qe0pzbE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L137)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

let's remove `less` now that we have `cmp`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 160 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgNuODu3cpV26wVvFXH:-OgXQI2BBTWIUYZET2bL:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c9f1d638bffa95896acc4e10368d2f0b277f6d/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L160)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

do we need this hash?
</blockquote></details>

Removed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAeRik8TBkK75sJEVD:-OgXYFAV5cE8ZCVrkDHs:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L238)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

comment is stale
</blockquote></details>

Changed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 263 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAeVaC3GL4kCdnE8s-:-OgXXq1fCOXP1RSNMPt6:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/85eac7be5ff7526f8388750ef1f17b610db5e708/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L263)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

comment is stale
</blockquote></details>

Changed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships` line 207 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgYIPVD3IodFoO45AnJ:-OgYNmW023sCFftUGRKH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93b27579e52454aaedf9d46443ed7a7f3f2b8940/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestNormalizedVoterAllRelationships#L207)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: this is no longer an "error case"
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAcssy7dmLVvg38q5y:-OgXUycB-JN0dJSr6ahP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L31)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The comment says :3, but here it is :1
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAd1Rm9E7Ct2ErxDZk:-OgXUwm17LBCs-eb1Amo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see any narrowing of the replica constraints in the output.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159210#-OgAd9pCCN-Is1lKc1MS:-OgXUyAoDU9p9TBd03CR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ca8f7b29578a7c4c0ee387b9707a262e1865863/pkg/kv/kvserver/allocator/mmaprototype/testdata/TestDedupConstraints#L92)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This shows :1 which contradicts the comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
